### PR TITLE
Sample gradient planner

### PR DESCRIPTION
**Pull request by thowell**
Experimental sampling-based planner:

First, sample $2 n_p$ policies:

$$
\begin{align*}
     \theta^{(i)} &= \theta + s e_i, \quad i = 1, \dots, n_p \\
     \theta^{(i + n_p)} &= \theta - s e_i, \quad i = 1, \dots, n_p
\end{align*}
$$

where $\theta \in \mathbf{R}^{n_p}$ is the nominal set of parameters, $s \in \mathbf{R}_{++}$ is a scaling control spread of samples, and $e_i \in \mathbf{R}^{n_p}$ is the $i$-th basis vector.

Second, construct approximate gradient and (diagonal) Hessian using finite differencing:

$$
\begin{align*}
     g^{(i)} &= \frac{R(\theta^{(i)}) - R(\theta^{(i + n_p)})}{2 s}\\
     H^{(i, i)} &= \frac{R(\theta^{(i)}) - 2 R(\theta) + R(\theta^{(i + n_p)})}{s^2}
\end{align*}
$$

where $g \in \mathbf{R}^{n_p}$, $H \in \mathbf{S}_{++}^{n_p}$, and $R : \mathbf{R}^{n_p} \rightarrow \mathbf{R}$ is the return of the policy.

The Cauchy point:

$$\theta^{(\text{C})} = \theta - \frac{g^T g}{g^T H g} \cdot g$$

and approximate Newton point ($H$ is diagonal):

$$ \theta^{(\text{N})} = \theta - H^{-1} g$$

are computed.

These points are used to construct another set of policies:

$$\phi^{(t)} = (1 - t) \cdot \theta^{(\text{C})} + t \cdot \theta^{(\text{N})}, \quad t \in [0, 1]$$

Finally, the new policy is the best among the perturbed policies $\theta^{(i)}$ and line-search policies $\phi^{(t)}$:

$$ \theta ={\text{argmin }} \{R(\theta), R(\theta^{(0)}), \dots, R(\theta^{(2 n_p)}), R(\phi^{(0)}), \dots, R(\phi^{(1)})\}$$

Importantly, in the case where perturbed parameters violate bounds, [alternative perturbations and finite-differencing schemes are employed](https://github.com/thowell/mujoco_mpc/blob/7abe46d4c7da98a6545103f64df6994fe41b75c0/mjpc/planners/sample_gradient/planner.cc#L382).

Ideas from [dog leg method](https://en.wikipedia.org/wiki/Powell%27s_dog_leg_method).

### CODE REVIEW

**Review by yuvaltassa — COMMENTED** [review]
Other than what i think is a bug in the gradient formula at the bounds this looks good.

Is it easy to add a color visualization (say, to the trace color) that will show us whether the winning traj came from the initial +- rollouts or from the Cauchy-Newton leg?

File: mjpc/planners/sample_gradient/planner.cc [hunk 1]
```diff
@@ -0,0 +1,689 @@
...
+  double scale2 = scale * scale;
+  double rc = (trajectory[2 * num_parameters].total_return - rm) / rstd; // nominal return
+
+  for (int i = 0; i < num_parameters; i++) {
+    double rp = (trajectory[i].total_return - rm) / rstd;
+    double rn = (trajectory[i + num_parameters].total_return - rm) / rstd;
+    if (parameter_status[i] == kParameterLower) {
+      // forward difference
+      gradient[i] = (-3 * rc + 4 * rp - rn) / (2 * scale);
+      hessian[i] = (rc - 2 * rp + rn) / scale2;
```
[the remark this answers is not in the document]
Comment (thowell) [reply mjpc/planners/sample_gradient/planner.cc]: Updated.

**Comment by thowell** [comment]
It's more challenging to change the winner color because `Agent` performs this visualization and `Planner` has no access. For now, I have modified the plot to step between 3 levels to indicated: perturb (-6), nominal (0), or gradient (6) winner.

image

File: mjpc/planners/sample_gradient/planner.cc (hunk 1 above)
[the remark this answers is not in the document]
Comment (thowell) [reply mjpc/planners/sample_gradient/planner.cc]: I added a reference to the difference schemes used and improved the comments.

**Pull request merged**